README.md
# SingleWireSerial

[![License: LGPL v3](https://img.shields.io/badge/License-LGPLv3-blue.svg)](https://www.gnu.org/licenses/lgpl-3.0)
[![Installation instructions](https://www.ardu-badge.com/badge/SingleWireSerial.svg?)](https://www.ardu-badge.com/SingleWireSerial)
[![Commits since latest](https://img.shields.io/github/commits-since/felias-fogg/SingleWireSerial/latest)](https://github.com/felias-fogg/SingleWireSerial/commits/master)
[![Build Status](https://github.com/felias-fogg/SingleWireSerial/workflows/LibraryBuild/badge.svg)](https://github.com/felias-fogg/SingleWireSerial/actions)
![Hit Counter](https://visitor-badge.laobi.icu/badge?page_id=felias-fogg_SingleWireSerial)


A software serial library that uses only a single wire for half-duplex serial communication. It makes use of the input capture and output compare match features of the 16-bit timers in AVR MCUs and is for this reason highly accurate and reliable up to 250 kbps. So it can be used, for instance, for debugWIRE communication. Furthermore,  you can also use it as a two wire library.




## Description

The SingleWireSerial library uses the *input capture* and the *output compare match* features of the 16-bit timers in AVR MCUs. Basically, you can time stamp falling or rising edges with the input capture feature and can accurately time intervals with the output compare match feature. This allows for a much more accurate way of interpreting an asynchronous signal than using wait loops, provided you want to support bit rates that are reconfigurable during runtime.

However, there is a price to pay. First of all, Timer 1, the 16-bit timer on the ATmega328P, is used for this purpose and for this reason, some of the PWM pins cannot be used for analog output (Arduino pins 9 and 10 on the ATmega328 boards). Second, the Servo library cannot be used, because it uses Timer 1. Third, there is a fixed input pin, namely Arduino pin 8 on ATmega328 boards. Fourth, since there is only one pin that can be used for input capture, you cannot have multiple instances of the SingleWireSerial class. 

There exist also support for running the library on other AVR MCUs than the ATmega328P (see `SingleWireSerial_config.h`). It should work for all MCUs which have a 16-bit timer, the input capture feature and the corresponding timer flag register among the first 32 I/O registers, but it is untested. It compiles for the following MCUs:

- ATmegaX8(P)
- ATmega32U4
- ATmega1284(P)
- ATmega2560
- ATtinyX4
- ATtinyX41
- ATtinyX7
- ATtinyX8
- ATtiny828


## Installation

Just copy the directory into the library directory of your Arduino installation.

## Usage

Usage of this library is very similar to the usage of the [`SoftwareSerial`](https://www.arduino.cc/en/Reference/softwareSerial) library. For this reason, only the differences to it are described here.

The main difference is that the constructor method has only one optional argument. This argument, called `twoWire`, if set to `true`, enables one to use an ordinary two-wire interface. The TX-line for an ATmega328 is in this case Arduino pin 9. For other MCUs, you find the RX-pin and TX-pin in `SingleWireSerial_config.h`. Further, the methods [`listen()`](https://www.arduino.cc/en/Reference/SoftwareSerialListen) and [`isListening()`](https://www.arduino.cc/en/Reference/SoftwareSerialIsListening) do not exist because one can define only one instance of the `SingleWireSerial` class.

## Limits

At 125 kbps at 16 MHz, timing is tight. If the millis timer interrupt is enabled, then, in the *worst case*, [**6.6 µs** are spent in the millis interrupt](https://hinterm-ziel.de/index.php/2021/10/20/timing-the-timekeeper/). In the worst case, it takes **3.8 µs** to set up the machinery in the receiving ISR after a falling edge. The critical point is at **12 µs** after a falling edge, at which we want to sample. So, there is no problem at 125 kbps provided the *only* interrupt is the millis interrupt.

At higher communication rates, e.g., at 250 kbps, the millis interrupt needs to be disabled when one wants to receive bytes reliably. And this goes for other interrupts without saying.

On the lower limit, it should be possible to go as low as 40 bps. However, note that all interrupts are blocked when writing a byte and when receiving a byte.

## Background

If you want to learn more about the background of the library, you might want read this [blog post](https://hinterm-ziel.de/index.php/2021/10/30/one-line-only/).


changelog.txt
1.2.0 - added _finishSendingEarly as a protected member that can be used
        in order to return before the end of the stop bit when sinding.
        This turned out to be necessary for dwSerial (the debugWIRE
        communcation class) because after changing the
        communication speed, you have to recalibrate communication speed
        which cannot be done by using the usual IRQ driven input.
      - added another possible prescaler (256) for ultra-slow bps down to 20
        bps
1.1.7 moved "enable input capture interrupt" to the start of the stop
      bit when writing a byte; this way, we do not miss the start bit
      of a following input byte!
1.1.6 changed attribute definitions from private to protected
      added TOV to the compile-time constants in SingleWireSerial_config.h
      use prescaler of 64 for all communication speeds lower than 4000
      bps, meaning that now we can deal with down to 40 bps
      GitHub actions are used to compile the examples when changes are
      made
      made some adjustments in SingleWireSerial_config.h
1.1.5 changed default timer for those MCUs that do not support input
      capture in their Timer1 modules
      added some code for Mega2560 & 1280 so that they can read from
      the respective input port (which cannot be addressed by IN and OUT)
1.1.4 removed the last change since it is impossible to parametrize it
      for different MCUs
1.1.3 fixed the problem that the sending routine starts before the
      stop bit of the receiving byte is fully received, which led to
      problems with debugWIRE
      handle_interrupt is now a private method (as it should!)
1.1.2 Corrected an annoying typo in the write method that led to the
      problem that the single-wire mode did not work anymore, modified
      the example, and declared setRxIntMsk to be protected so that
      derived classes can use it
1.1.1 Rearranged the initial instructions.
1.1.0 Complete redesign of receiving ISR. It is now all inline assembler,
      it uses only 4 registers + SREG and employs OCRA for bit timing. Is now millis
      interrupt safe with 125,000 bps and still has one bit time (4
      microseconds) between bytes when working at 250,000 bps with the millis
      interrupt disabled.
1.0.0 first publication
0.9.2 handle_interrupt redesigned (is faster now and terminates earlier)
0.9.1b curr deleted
0.9.1a setRxIntMsk static / recv eliminated
0.9.1 Initial commit
      


src/SingleWireSerial.h
/*
  SingleWireSerial.h - A software serial library that uses only
  one wire to connect two systems in half-duplex mode. In addition,
  it uses the input capture feature and output 
  compare match feature of timer 1 in order to support high bit rates.
  It is loosely based on SoftwareSerial, but uses a completely
  different method for reading and writing. 

  Since one usually has only one input capture pin, it does not make
  sense to have more than one instance. 

  This library is free software; you can redistribute it and/or
  modify it under the terms of the GNU Lesser General Public
  License as published by the Free Software Foundation; either
  version 2.1 of the License, or (at your option) any later version.
  
  This library is distributed in the hope that it will be useful,
  but WITHOUT ANY WARRANTY; without even the implied warranty of
  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
  Lesser General Public License for more details.
  
  You should have received a copy of the GNU Lesser General Public
  License along with this library; if not, write to the Free Software
  Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA  02110-1301  USA

*/

#ifndef SingleWireSerial_h
#define SingleWireSerial_h
#include <SingleWireSerial_config.h>
#include <inttypes.h>
#include <Stream.h>

/******************************************************************************
* Definitions
******************************************************************************/

#ifndef _SS_MAX_RX_BUFF
#define _SS_MAX_RX_BUFF 64 // RX buffer size
#endif

class SingleWireSerial : public Stream
{
private:
  // only 1 object possible, so everything is static
  static bool _twoWire;
  static bool _waitBeforeSending;
  static bool _buffer_overflow;
  static uint16_t _oneAndAHalfBitDelay;
  static uint16_t _bitDelay;
  static uint16_t _endOfByte;
  static uint8_t _setICfalling, _setICrising, _setCTC;
  static uint8_t _receive_buffer[_SS_MAX_RX_BUFF]; 
  static volatile uint8_t _receive_buffer_tail;
  static volatile uint8_t _receive_buffer_head;


protected:
  static void setRxIntMsk(bool enable);

public:
  // public methods
  SingleWireSerial(bool twoWire = false);
  ~SingleWireSerial();
  void begin(long speed);
  void end();
  bool overflow(); 
  int peek();

  virtual size_t write(uint8_t data);
  virtual int read();
  virtual int available();
  virtual void flush();
  operator bool() { return true; }

  // public only for easy access by interrupt handler
  static inline void handle_interrupt() __attribute__((__always_inline__));
};

#endif


src/SingleWireSerial.cpp
/*
  SingleWireSerial.cpp -  A software serial library that uses only
  one wire to connect two systems in half-duplex mode. In addition,
  it uses the input capture feature and output 
  compare match feature of timer 1 in order to support high bit rates.
  It is loosely based on SoftwareSerial, but uses a completely
  different method for reading and writing. 

  Since one usually has only one input capture pin, it does not make
  sense to have more than one instance. 

  This library is free software; you can redistribute it and/or
  modify it under the terms of the GNU Lesser General Public
  License as published by the Free Software Foundation; either
  version 2.1 of the License, or (at your option) any later version.
  
  This library is distributed in the hope that it will be useful,
  but WITHOUT ANY WARRANTY; without even the implied warranty of
  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
  Lesser General Public License for more details.
  
  You should have received a copy of the GNU Lesser General Public
  License along with this library; if not, write to the Free Software
  Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA  02110-1301  USA

*/

// When _DEBUG  == 1, then some bit pulses are sent to all PORTC-pins (i.e.,
// analog and I2C pins. The on/off toggle needs 2 cycles, but may, of course,
// disturb the timing a bit. When _LOGDEBUG == 1, some info is printed
// using the ordinary Serial connection (if it is open).
#define _DEBUG 0
#define _LOGDEBUG 0
#define _FASTIRQ 1 // This is the current version, set it to 1!
// 
// Includes
// 
#include <avr/interrupt.h>
#include <avr/pgmspace.h>
#include <Arduino.h>
#include "SingleWireSerial.h"
#include <util/delay_basic.h>

// Statics
// The used attribute is necessary, otherwise the compiler otimizes the variables away!
bool SingleWireSerial::_twoWire;
bool SingleWireSerial::_waitBeforeSending asm("_waitBeforeSending") __attribute__ ((used));
bool SingleWireSerial::_buffer_overflow asm("_buffer_overflow") __attribute__ ((used));
uint16_t SingleWireSerial::_bitDelay asm("_bitDelay") __attribute__ ((used)) = 0;
uint16_t SingleWireSerial::_oneAndAHalfBitDelay asm("_oneAndAHalfBitDelay")  __attribute__ ((used));
uint16_t SingleWireSerial::_endOfByte;

uint8_t SingleWireSerial::_setICfalling asm("_setICfalling") __attribute__ ((used));
uint8_t SingleWireSerial::_setICrising asm("_setICrising") __attribute__ ((used));
uint8_t SingleWireSerial::_setCTC asm("_setCTC") __attribute__ ((used));

uint8_t SingleWireSerial::_receive_buffer[_SS_MAX_RX_BUFF] asm("_receive_buffer")  __attribute__ ((used)); 
volatile uint8_t SingleWireSerial::_receive_buffer_tail asm("_receive_buffer_tail") = 0;
volatile uint8_t SingleWireSerial::_receive_buffer_head asm("_receive_buffer_head") = 0;

//
// Debugging
//
// This function generates a brief pulse
// for debugging or measuring on an oscilloscope.
// All in all, it takes 4 cycles to produce the pulse
#if _DEBUG
inline __attribute__ ((always_inline)) void DebugPulse(byte signal)
{
  PORTC |= signal;
  PORTC &= ~signal;
}
#else 
inline __attribute__ ((always_inline)) void DebugPulse(__attribute__ ((unused)) byte signal) {}
#endif

//
// Interrupt handling
//

/* static */
void SingleWireSerial::handle_interrupt() 
#if _FASTIRQ
{
  asm volatile(
	       // save registers
	       "push r24\n\t"
	       "lds r24,  %A[ICRaddr] ; load ICR low, now high byte is saved!\n\t"
#if _DEBUG
	       "sbi %[PORTCaddr], 0\n\t"
	       "cbi %[PORTCaddr], 0\n\t"
#endif
	       "push r25\n\t"
	       "lds r25,  %B[ICRaddr] ; load ICR high\n\t"
	       "push r30\n\t"
	       "push r31\n\t"
	       "in r30, __SREG__\n\t"
	       "push r30 ; save status register\n\t"
	       
	       // compute time since edge interrupt 
	       "lds r30,  %A[TCNTaddr] ; load TCNT low\n\t"
	       "lds r31,  %B[TCNTaddr] ; load TCNT high\n\t"
	       "sub r30, r24 ; TCNT - ICR: time since edge IRQ\n\t"
	       "sbc r31, r25\n\t"

	       // setup OCR for first 1.5 bit times
	       "lds r24, _oneAndAHalfBitDelay\n\t"
	       "lds r25, _oneAndAHalfBitDelay+1\n\t"
	       "sts %B[OCRAaddr], r25\n\t"
	       "sts %A[OCRAaddr], r24\n\t"

	       // load CTC setting, check for slowness, store TCNT and TCCRB, clear flag
	       "lds r24, _setCTC ; set counter to CTC operation\n\t"
	       "sbrs r24, %[SLOWCS] ; when slow bit is set, skip\n\t"
	       "adiw r30, %[STARTOFFSET] ; time that is unaccounted for\n\t"
	       "sts %B[TCNTaddr], r31 ; store back to TCNT\n\t"
	       "sts %A[TCNTaddr], r30\n\t"
	       "sts %[TCCRBaddr], r24 ; set CTC mode\n\t"
	       "sbi %[TIFRIOaddr], %[OCFAconst] ; clear flag OVF flag\n\t"
#if _DEBUG
	       "sbi %[PORTCaddr], 0\n\t"
	       "cbi %[PORTCaddr], 0\n\t"
#endif	       
	       
	       // compute & remember where to store input char
	       "lds r30, _receive_buffer_tail\n\t"
	       "mov r24, r30 ; remember tail for later\n\t"
	       "ldi r31, 0\n\t"
	       "subi r30, lo8(-(_receive_buffer))\n\t"
	       "sbci r31, hi8(-(_receive_buffer)) ; r30:r31 is now address of _receive_buffer[_receive_buffer_tail]\n\t"
	       "push 30 ; save for later \n\t"
	       "push 31\n\t"

	       // advance _receive_buffer_tail and check for overrun
	       "inc r24 ; increment tail\n\t"
	       "andi r24, %[BUFFMASK] ; do a simple modulo with a power of 2 for a number <= 256\n\t"
	       "lds r25, _receive_buffer_head\n\t"
	       "cp r24, r25 ; compare now with head ptr\n\t"
	       "breq _LOVF ; if equal, mark overflow\n\t"
	       "sts  _receive_buffer_tail, r24 ; store new tail\n\t"
	       "rjmp _LSTARTREAD ; now start to read bits\n\t"
	       "_LOVF: ldi r24, 0x01\n\t"
	       "sts _buffer_overflow, r24 ; overflow bit is set\n\t"

	       // initialize ch (r24) and bit counter (r25)
	       "_LSTARTREAD: ldi r25, 0x08\n\t"
	       "clr r24\n\t"

#if _DEBUG
	       "sbi %[PORTCaddr], 0\n\t"
	       "cbi %[PORTCaddr], 0\n\t"
#endif	       

	       // loop for first 7 bits
	       "_LBYTELOOP: subi r25, 0x01\n\t"
	       "breq _LASTBIT\n\t"

	       // wait for flag to be set
	       "clc ; clear carry bit - needed later for bit to read\n\t"
	       "_LWAIT1: sbis %[TIFRIOaddr], %[OCFAconst]\n\t"
	       "rjmp _LWAIT1\n\t"
	       
	       // read one bit
#if defined(__AVR_ATmega1280__) || defined(__AVR_ATmega2560__)
	       "lds r30, %[INPORTaddr]\n\t"
	       "sbrc r30, %[INPIN] ; skip if cleared\n\t"
#else
	       "sbic %[INPORT], %[INPIN] ; skip if cleared\n\t"
#endif
	       "sec ; set carry if bit is set\n\t"
	       "ror r24 ; rotate carry bit into byte\n\t"
#if _DEBUG
	       "sbi %[PORTCaddr], 2\n\t"
	       "cbi %[PORTCaddr], 2\n\t"
#endif
	       // setup for next bit
	       "lds r30, _bitDelay ; set OCRA now to _bitDelay\n\t"
	       "lds r31, _bitDelay+1\n\t"
	       "sts %B[OCRAaddr], r31\n\t"
	       "sts %A[OCRAaddr], r30\n\t"
	       "sbi %[TIFRIOaddr], %[OCFAconst] ; clear flag\n\t"
	       "rjmp _LBYTELOOP\n\t"

	       // now the last bit (r30:31 contains _bitDelay)
	       "_LASTBIT: sbiw r30, %[ENDOFFSET] ; subtract everything for what we do before the last read = 10 cyc\n\t"
	       "sts %B[OCRAaddr], r31\n\t"
	       "sts %A[OCRAaddr], r30\n\t"

      	       // set _waitBeforeSending to true
	       "ldi r30, 1\n\t"
	       "sts _waitBeforeSending, r30\n\t"

	       // wait for flag to be set
	       "clc ; clear carry bit - needed later for bit to read\n\t"
	       "_LWAIT2: sbis %[TIFRIOaddr], %[OCFAconst]\n\t"
	       "rjmp _LWAIT2\n\t"

	       // set TCCRB, clear ICF and restore bufptr
	       "lds r30, _setICfalling\n\t"
	       "sts %[TCCRBaddr], r30\n\t"
	       "pop r31 ; restore bufptr\n\t"
	       "pop r30\n\t"
	       "sbi %[TIFRIOaddr], %[ICFconst]\n\t"

	       // read last bit
#if defined(__AVR_ATmega1280__) || defined(__AVR_ATmega2560__)
	       "lds r25, %[INPORTaddr]\n\t"
	       "sbrc r25, %[INPIN] ; skip if cleared\n\t"
#else
	       "sbic %[INPORT], %[INPIN] ; skip if cleared\n\t"
#endif
	       "sec ; set carry if bit is set\n\t"
	       "ror r24 ; rotate carry bit into byte\n\t"
#if _DEBUG
	       "sbi %[PORTCaddr], 2\n\t"
	       "cbi %[PORTCaddr], 2\n\t"
#endif
	       // finish
	       "st z, r24 ; store read character\n\t"
	       "pop r31 ; status register\n\t"
	       "out __SREG__, r31\n\t"
	       "pop r31\n\t"
	       "pop r30\n\t"
	       "pop r25\n\t"
	       "pop r24\n\t"
#if _DEBUG
	       "sbi %[PORTCaddr], 1\n\t"
	       "cbi %[PORTCaddr], 1\n\t"
#endif
	       "reti ; done\n\t"
	       :
	       :
#if _DEBUG
	         [PORTCaddr] "I" (_SFR_IO_ADDR(PORTC)),
#endif
		 [ICRaddr] "M" (&ICR),
		 [TCNTaddr] "M" (&TCNT),
		 [STARTOFFSET] "I" (35),
		 [BUFFMASK] "M" (_SS_MAX_RX_BUFF-1),
		 [OCRAaddr] "M" (&OCRA), 
		 [TCCRBaddr] "M" (&TCCRB),
	         [TIFRIOaddr] "M" (_SFR_IO_ADDR(TIFR)),
		 [OCFAconst] "M" (OCFA),
#if defined(__AVR_ATmega1280__) || defined(__AVR_ATmega2560__)
		 [INPORTaddr] "n" (&ICPIN),
#else
		 [INPORT] "I" (_SFR_IO_ADDR(ICPIN)),
#endif
		 [INPIN] "I"  (ICBIT),
		 [ENDOFFSET] "M" (10),
		 [ICFconst] "M" (ICF),
	         [SLOWCS] "M" (CS0));
}
#else // not _FASTIRQ
{
  uint8_t ch, level;
  uint16_t elapsed,  next; // read input capture register
  uint16_t start;
  byte * bufptr;

  // one only has to save SREG, r0-r1, r18-r27, 30-31
  // since only these are the registers that are
  // used in a C-routine without restoring them
  asm volatile("push r22 \n\t"
	       "push r23 \n\t"
	       "lds r22, _setICrising ; load code for raising edge \n\t"
	       "sts %[TCCRBaddr], r22 ; write to control register \n\t"
	       "lds  r22, %A[ICRaddr] ; load ICR low byte\n\t"
	       "lds  r23, %B[ICRaddr] ; load ICR high byte\n\t"
#if _DEBUG
	       "sbi %[PORTCaddr], 0\n\t"
	       "cbi %[PORTCaddr], 0\n\t"
#endif
	       "push r0 ; save temp reg\n\t"
	       "in r0, __SREG__ ; save status reg\n\t"
	       "push r0\n\t"
	       "push r1\n\t"
	       "eor r1, r1 ; clear r1\n\t"
	       "push r18\n\t"
	       "push r19\n\t"
	       "push r20\n\t"
	       "push r21\n\t"
	       "push r24\n\t"
	       "push r25\n\t"
	       "push r26\n\t"
	       "push r27\n\t"
	       "push r30\n\t"
	       "push r31\n\t"
	       "movw %[START], r22\n\t"
	       : [START] "=d" (start)
	       : [ICRaddr] "M" (&ICR),
		 [TCCRBaddr] "M" (&TCCRB),
		 [EDGEUP] "M" (_BV(ICES)),
#if _DEBUG
		 [PORTCaddr] "I" (_SFR_IO_ADDR(PORTC))
#endif
	       );
  setRxIntMsk(false); // disable the ICR interrupts
  ch = 0;
  level = 0;
  next = _oneAndAHalfBitDelay;

  // tail points to where byte goes
  bufptr = &_receive_buffer[_receive_buffer_tail];
  // if buffer full, set the overflow flag
  uint8_t nextix = (_receive_buffer_tail + 1) % _SS_MAX_RX_BUFF;
  if (nextix != _receive_buffer_head)
    {
      // that is where next byte shall go
      _receive_buffer_tail = nextix;
    } 
  else 
    {
      DebugPulse(0x08);
      _buffer_overflow = true;
    }
  DebugPulse(0x01);
  
  while (next <= _endOfByte) {
    if (TIFR & _BV(ICF)) { // capture flag has been set
      DebugPulse(0x04);
      TIFR |= _BV(ICF); // clear flag
      TCCRB ^= _BV(ICES); // toggle edge detector;
      level ^= 0x80;
    }
    elapsed = TCNT - start;  // 16 bit unsigned arithmetic gives correct duration
    if (elapsed > next) { 
      ch >>=1;
      ch |= level;
      next = next + _bitDelay;
    }
  }
  _waitBeforeSending = 1;
  *bufptr = ch; // save new byte
  TCCRB &= ~_BV(ICES); // set edge detector to falling edge

  setRxIntMsk(true); // and enable input capture interrupt again
  DebugPulse(0x02);
  asm volatile(
	       "pop r31\n\t"
	       "pop r30\n\t"
	       "pop r27\n\t"
	       "pop r26\n\t"
	       "pop r25\n\t"
	       "pop r24\n\t"
	       "pop r21\n\t"
	       "pop r20\n\t"
	       "pop r19\n\t"
	       "pop r18\n\t"
	       "pop r1\n\t"
	       "pop r0\n\t"
	       "out __SREG__, r0\n\t"
	       "pop r0\n\t"
	       "pop r23\n\t"
	       "pop r22 \n\t"
#if _DEBUG
	       "sbi %[PORTCaddr], 1\n\t"
	       "cbi %[PORTCaddr], 1\n\t"
#endif
	       "reti"
	       :
	       :
#if _DEBUG
	       [PORTCaddr] "I" (_SFR_IO_ADDR(PORTC))
#endif
	       );
}
#endif // not _FASTIRQ

ISR(TIMER_CAPT_vect, ISR_NAKED)
{
  SingleWireSerial::handle_interrupt();
}

//
// Constructor
//
SingleWireSerial::SingleWireSerial(bool twoWire)
{
  _twoWire = twoWire;
  _buffer_overflow = false;
  pinMode(ICArduinoPin, INPUT);
  if (twoWire) {
    digitalWrite(OCArduinoPin, HIGH);
    pinMode(OCArduinoPin, OUTPUT);
  }
}

//
// Destructor
//
SingleWireSerial::~SingleWireSerial()
{
  end();
}

void  SingleWireSerial::setRxIntMsk(bool enable)
{
  if (enable) {
    TCCRB = _setICfalling; // look for falling edge of start bit
    TIFR |= _BV(ICF); // clear input capture flag
    TIMSK |= _BV(ICIE); // enable interrupt
  } else
    TIMSK &= ~_BV(ICIE); // disable input capture interrupt
}

//
// Public methods
//

void SingleWireSerial::begin(long speed)
{
  
  // Precalculate the various delays
  uint32_t bit_delay100 = (F_CPU*100 / speed);
  uint8_t prescaler;

  // init _waitBeforeSending
  _waitBeforeSending = true;
  
  // clear read buffer
   _receive_buffer_tail = _receive_buffer_head;
   _buffer_overflow = false;

  if (bit_delay100 > 400000UL) {
    bit_delay100 = bit_delay100/64;
    prescaler = _BV(CS1)|_BV(CS0); // prescaler = 64
  } else {
    prescaler = _BV(CS0); // prescaler = 1
  }
  _setICfalling = _BV(ICNC) | prescaler;
  _setICrising = _setICfalling | _BV(ICES);
  _setCTC = _BV(WGM2) | prescaler;

  _bitDelay = (bit_delay100+50)/100; // bit delay time in timer1 ticks
  _oneAndAHalfBitDelay = (bit_delay100+bit_delay100/2+50)/100; // delay until first sample time point
  _endOfByte = _oneAndAHalfBitDelay + (7*_bitDelay); // last sample timepoint

#if _LOGDEBUG
  Serial.print(F("bit_delay100="));
  Serial.println(bit_delay100);
  Serial.print(F("prescaler="));
  Serial.println(prescaler);
  Serial.print(F("_bitDelay="));
  Serial.println(_bitDelay);
  Serial.print(F("_oneAndAHalfBitDelay="));
  Serial.println(_oneAndAHalfBitDelay);
  Serial.print(F("_endOfByte="));
  Serial.println(_endOfByte);
  Serial.print(F("_setCTC="));
  Serial.println(_setCTC,BIN);
#endif
  TCCRA = 0;
  TCCRC = 0;
  setRxIntMsk(true);
#if _DEBUG
  DDRC |= 0x3F;
#endif
}

void SingleWireSerial::end()
{
  _receive_buffer_tail = _receive_buffer_head;
  setRxIntMsk(false);
}


// Read data from buffer
int SingleWireSerial::read()
{
  // Empty buffer?
  if (_receive_buffer_head == _receive_buffer_tail)
    return -1;

  // Read from "head"
  uint8_t d = _receive_buffer[_receive_buffer_head]; // grab next byte
  _receive_buffer_head = (_receive_buffer_head + 1) % _SS_MAX_RX_BUFF;
  return d;
}

int SingleWireSerial::available()
{
  return ((unsigned int)(_SS_MAX_RX_BUFF + _receive_buffer_head - _receive_buffer_tail)) % _SS_MAX_RX_BUFF;
}

size_t SingleWireSerial::write(uint8_t data)
{
  uint8_t oldSREG = SREG;

  setRxIntMsk(false);
  TCCRA = 0;
  TCCRC = 0;
  TCCRB = _setCTC;
  OCRA = _bitDelay-1;

  cli(); // interrupts off

  TCNT = 0;
  TIFR |= _BV(OCFA);
  DebugPulse(0x01);

  if (_waitBeforeSending) { // wait here, because we do not wait to the end of the stop bit when reading
    _waitBeforeSending = false;
    OCRA = _bitDelay << 1; // two bit times
    while (!(TIFR & _BV(OCFA)));
    OCRA = _bitDelay-1;
    TCNT = 0;
    TIFR |= _BV(OCFA);
  }
  
  if (!_twoWire) {
    TCNT = 0;
    ICDDR |= _BV(ICBIT);   // startbit
    DebugPulse(0x04);
    for (uint8_t i = 8; i > 0; --i) {
      while (!(TIFR & _BV(OCFA)));
      if (data & 1) {
	ICDDR &= ~_BV(ICBIT); // make output high-impedance
	DebugPulse(0x02);
      } else {
	ICDDR |= _BV(ICBIT); // pull-down
	DebugPulse(0x04);
      }
      TIFR |= _BV(OCFA);
      data >>= 1;
    } while (!(TIFR & _BV(OCFA)));
    ICDDR &= ~_BV(ICBIT); // make output again high-impedance for stop bit
    DebugPulse(0x02);
  } else { // twoWire!
    TCNT = 0;
    OCPORT &= ~_BV(OCBIT);  // startbit
    for (uint8_t i = 8; i > 0; --i) {
      while (!(TIFR & _BV(OCFA)));
      if (data & 1)
	OCPORT |= _BV(OCBIT); // make output high
      else
	OCPORT &= ~_BV(OCBIT); // make output low
      TIFR |= _BV(OCFA);
      data >>= 1;
    } while (!(TIFR & _BV(OCFA)));
    OCPORT |= _BV(OCBIT); // make output again high for stop bit
  }
  TIFR |= _BV(OCFA); // clear overflow flag

  SREG = oldSREG; // enable interrupts again
  setRxIntMsk(true); //enable input capture input interrupts again

  while (!(TIFR & _BV(OCFA))); // wait for stop bit to finish
  DebugPulse(0x01);
  return 1;
}

void SingleWireSerial::flush()
{
  // There is no tx buffering, simply return
}

int SingleWireSerial::peek()
{
  // Empty buffer?
  if (_receive_buffer_head == _receive_buffer_tail)
    return -1;

  // Read from "head"
  return _receive_buffer[_receive_buffer_head];
}

bool SingleWireSerial::overflow()
{
  bool ret = _buffer_overflow;
  if (ret)
    _buffer_overflow = false;
  return ret;
}


